watchdog.go

```diff
@@ -501,3 +506,31 @@ func wdrecover() {
 		}
 	}
 }
+
+// RegisterNotifee registers a function that will be called when a GC has happened.
+// The unregister function returned can be used to unregister this notifee.
+func RegisterNotifee(f func()) (unregister func()) {
+	notifeeMutex.Lock()
+	defer notifeeMutex.Unlock()
+
+	id := notifeeMaxIndex
+	notifeeMaxIndex++
+	notifees[id] = f
+
+	return func() {
+		notifeeMutex.Lock()
+		delete(notifees, id)
+		notifeeMutex.Unlock()
+	}
+}
+
+func notifyGC() {
+	if NotifyGC != nil {
+		NotifyGC()
+	}
+	notifeeMutex.Lock()
+	defer notifeeMutex.Unlock()
+	for _, f := range notifees {
+		f()
+	}
```

Comment:
I just realised that map iteration won't preserve the call order of registration, and will effectively make call order random. I think deterministic order is an important property to have. Could you turn this into a slice, and have the unregister function iterate over the slice and delete by equality comparison? Sorry for the late realisation!

Reply:
You can't compare functions in Go (that was actually my first attempt at implementing this).
We can delete from the slice, as long as we're not optimizing for a large number of notifees.

Reply:
Actually, deleting from the slice doesn't work either, as the index (which would be captured in `unregister`) is not stable then.